src/statsmodels/formula.jl

```diff
@@ -234,34 +240,149 @@ function dropUnusedLevels!(da::PooledDataArray)
+                ##is_redundant = any([dropped == enc for enc in encountered_columns])
+
+                if !is_redundant
+                    to_promote[i_eterm, i_term] = true
+                    push!(encountered_columns, dropped)
+                end
+
+            end
+            ## once we've checked all the eterms in this term, add it to the list
+            ## of encountered terms/columns
+        end
+        push!(encountered_columns, trms.factors[:, i_term])
+    end
+
+    return to_promote
+    
+end
+
+## Goal here is to allow specification of _either_ a "naked" contrast type,
+## or an instantiated contrast object itself.  This might be achieved in a more
+## julian way by overloading call for c::AbstractContrasts to just return c.
+evaluateContrasts(c::AbstractContrasts, col::AbstractDataVector) = ContrastsMatrix(c, col)
+evaluateContrasts{C <: AbstractContrasts}(c::Type{C}, col::AbstractDataVector) = ContrastsMatrix(c(), col)
+evaluateContrasts(c::ContrastsMatrix, col::AbstractDataVector) = c
+
+needsContrasts(::PooledDataArray) = true
+needsContrasts(::Any) = false
+
+function ModelFrame(trms::Terms, d::AbstractDataFrame;
+                    contrasts::Dict = Dict())
     df, msng = na_omit(DataFrame(map(x -> d[x], trms.eterms)))
     names!(df, convert(Vector{Symbol}, map(string, trms.eterms)))
     for c in eachcol(df) dropUnusedLevels!(c[2]) end
-    ModelFrame(df, trms, msng)
+
+    ## Set up contrasts: 
+    ## Combine actual DF columns and contrast types if necessary to compute the
+    ## actual contrasts matrices, levels, and term names (using TreatmentContrasts
+    ## as the default)
+    evaledContrasts = Dict()
+    for (term, col) in eachcol(df)
+        needsContrasts(col) || continue
+        evaledContrasts[term] = evaluateContrasts(haskey(contrasts, term) ?
+                                                  contrasts[term] :
+                                                  TreatmentContrasts,
+                                                  col)
+    end
+
+    ## Check whether or not
+    non_redundants = check_non_redundancy(trms, df)
+
+    ModelFrame(df, trms, msng, evaledContrasts, non_redundants)
 end
 
-ModelFrame(f::Formula, d::AbstractDataFrame) = ModelFrame(Terms(f), d)
-ModelFrame(ex::Expr, d::AbstractDataFrame) = ModelFrame(Formula(ex), d)
+ModelFrame(f::Formula, d::AbstractDataFrame; kwargs...) = ModelFrame(Terms(f), d; kwargs...)
+ModelFrame(ex::Expr, d::AbstractDataFrame; kwargs...) = ModelFrame(Formula(ex), d; kwargs...)
+
+## modify contrasts in place
+function contrast!(mf::ModelFrame, new_contrasts::Dict)
+    new_contrasts = [ col => evaluateContrasts(contr, mf.df[col])
+                      for (col, contr) in filter((k,v)->haskey(mf.df, k), new_contrasts) ]
+                      
+    mf.contrasts = merge(mf.contrasts, new_contrasts)
+    return mf
+end
+contrast!(mf::ModelFrame; kwargs...) = contrast!(mf, Dict(kwargs))
```

Comment:
Maybe a more natural API would be that the method above *replaces* the contrasts, leaving unspecified columns use the default (treatment) contrasts, while this method would only modify the specified ones? Else, you wouldn't have an easy way of resetting contrasts. Not sure.